extensions/gcp/processors/PutGcsObject.cpp

```diff
@@ -0,0 +1,304 @@
+    }
+  }
+
+ private:
+  std::string bucket_;
+  std::string key_;
+  gcs::Client& client_;
+
+  gcs::MD5HashValue hash_value_;
+  gcs::Crc32cChecksumValue crc32c_checksum_;
+  gcs::EncryptionKey encryption_key_;
+  gcs::PredefinedAcl predefined_acl_;
+  gcs::ContentType content_type_;
+  gcs::IfGenerationMatch if_generation_match_;
+
+  google::cloud::StatusOr<gcs::ObjectMetadata> result_;
+};
+
+[[nodiscard]] std::optional<std::string> getContentType(const core::ProcessContext& context, const core::FlowFile& flow_file) {
+  return context.getProperty(PutGcsObject::ContentType) | utils::orElse([&flow_file] {return flow_file.getAttribute("mime.type");});
+}
+
+std::shared_ptr<google::cloud::storage::oauth2::Credentials> getCredentials(core::ProcessContext& context) {
+  std::string service_name;
+  if (context.getProperty(PutGcsObject::GCPCredentials.getName(), service_name) && !IsNullOrEmpty(service_name))
+    return std::dynamic_pointer_cast<const GcpCredentialsControllerService>(context.getControllerService(service_name))->getCredentials();
+  return nullptr;
+}
+}  // namespace
+
+
+void PutGcsObject::initialize() {
+  setSupportedProperties({GCPCredentials,
+                          Bucket,
+                          ObjectName,
+                          NumberOfRetries,
+                          ContentType,
+                          MD5HashLocation,
+                          Crc32cChecksumLocation,
+                          EncryptionKey,
+                          ObjectACL,
+                          OverwriteObject,
+                          EndpointOverrideURL});
+  setSupportedRelationships({Success, Failure});
+}
+
+gcs::Client PutGcsObject::getClient(const gcs::ClientOptions& options) const {
+  return gcs::Client(options, *retry_policy_);
+}
+
+
+void PutGcsObject::onSchedule(const std::shared_ptr<core::ProcessContext>& context, const std::shared_ptr<core::ProcessSessionFactory>&) {
+  gsl_Expects(context);
+  if (auto number_of_retries = context->getProperty<uint64_t>(NumberOfRetries)) {
+    retry_policy_ = std::make_shared<google::cloud::storage::LimitedErrorCountRetryPolicy>(*number_of_retries);
+  }
+  if (auto encryption_key = context->getProperty(EncryptionKey)) {
+    try {
+      encryption_key_ = gcs::EncryptionKey::FromBase64Key(*encryption_key);
+    } catch (const google::cloud::RuntimeStatusError&) {
+      throw minifi::Exception(ExceptionType::PROCESSOR_EXCEPTION, EncryptionKey.getName() + " is not in base64: " + *encryption_key);
+    }
+  }
+  gcp_credentials_ = getCredentials(*context);
+}
+
+void PutGcsObject::onTrigger(const std::shared_ptr<core::ProcessContext>& context, const std::shared_ptr<core::ProcessSession>& session) {
+  gsl_Expects(context && session);
+
+  if (!gcp_credentials_) {
+    logger_->log_error("Invalid or missing credentials from Google Cloud Platform Credentials Controller Service");
+    context->yield();
+    return;
+  }
```

Comment:
we could move this check to `onSchedule` and treat the "non-empty `gcp_credentials_`" as a precondition for `onTrigger`

Reply:
Good idea, changed it in https://github.com/apache/nifi-minifi-cpp/pull/1268/commits/57404afcf541daa1780365e054a555dcd8093c3e#diff-8d49e2bfb2386efa4553c51d312468939c64efcd37b38f8b0e5871fbdade4a2aR229-R234